client/optimizing.go

```diff
@@ -0,0 +1,146 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/log"
+)
+
+// defaultTTL is the time a RTT sample will live before it is tested again.
+const defaultTTL = time.Minute * 5
+
+// NewOptimizingClient creates a drand client that measures the speed of clients
+// and uses the fastest one. Speeds are measured as the optimizing client is
+// used, so there is no background routine and no additional calls to other
+// clients.
+//
+// A speed measurement lives for a certain period before it is reset. The next
+// call to `.Get` will then attempt to use a reset client even if it was
+// previously considered to be slower. Pass a value > 0 for `rttTTL` to
+// customize the period - it defaults to 5 minutes.
+//
+// If a client fails to return a value it'll be attempted from the next fastest
+// client. Failed clients are given a large RTT and are moved to the back of the
+// list. This RTT expires as usual and allows failed clients to be re-introduced
+// and become the fastest again. i.e. failed clients are not taken out of the
+// list, just considered to be very slow.
+func NewOptimizingClient(clients []Client, chainInfo *chain.Info, rttTTL time.Duration) (Client, error) {
+	if len(clients) == 0 {
+		return nil, errors.New("missing clients")
+	}
+	if chainInfo == nil {
+		return nil, errors.New("missing chain info")
+	}
+	stats := make([]*clientStat, len(clients))
+	now := time.Now()
+	for i, c := range clients {
+		stats[i] = &clientStat{client: c, rtt: 0, ts: now}
+	}
+	if rttTTL <= 0 {
+		rttTTL = defaultTTL
+	}
+	return &optimizingClient{stats: stats, rttTTL: rttTTL, chainInfo: chainInfo, log: log.DefaultLogger}, nil
+}
+
+type optimizingClient struct {
+	sync.RWMutex
+	stats     []*clientStat
+	rttTTL    time.Duration
+	chainInfo *chain.Info
+	log       log.Logger
+}
+
+type clientStat struct {
+	client Client
+	rtt    time.Duration
+	ts     time.Time
+}
+
+// SetLog configures the client log output
+func (oc *optimizingClient) SetLog(l log.Logger) {
+	oc.log = l
+}
+
+// Get returns a the randomness at `round` or an error.
+func (oc *optimizingClient) Get(ctx context.Context, round uint64) (res Result, err error) {
+	oc.RLock()
+	// take a copy of the current client stats so we iterate over a stable slice
+	currStats := make([]*clientStat, len(oc.stats))
+	copy(currStats, oc.stats)
+	oc.RUnlock()
+
+	nextStats := []*clientStat{}
+	for _, s := range currStats {
+		start := time.Now()
+		res, err = s.client.Get(ctx, round)
+		rtt := time.Now().Sub(start)
+
+		// client failure, set a large RTT so it is sent to the back of the list
+		if err != nil {
+			rtt = math.MaxInt64
+		}
+
+		nextStats = append(nextStats, &clientStat{s.client, rtt, start})
+
+		if err == nil {
+			break
+		}
+
+		// context deadline hit
+		if ctx.Err() != nil {
```

Comment:
if the downstream consumer does something like
```golang
ctx, cancel = context.WithTimeout(context.Background(), time.Second)
defer cancel()
client.Get(ctx, 0)
```

can easily attempt to get from the first client, time out on that one, and never try or re-order others.

Reply:
Not sure I understand? If the context is canceled then we don't want it to try others, and because it was canceled the stat is not useful and not worth recording.

Reply:
the worry is that if i've got a slow client at the front of the queue, and any sort of time bounds on the get calls, i may well never attempt / fail over to other clients, because i time-out before reaching them.

Reply:
Ah, ok I understand, but, its a problem with the existing prioritizing client also right?

Reply:
correct